packages/opentelemetry-exporter/src/index.js

```diff
@@ -0,0 +1,288 @@
+const INTERNAL_KIND = 'internal';
+/**
+ * https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto#L314
+ * Status Code
+ *
+ * Default
+ * UNSET = 0,
+ * The operation has been validated by an Application developer or
+ * Operator to have completed successfully.
+ * OK = 1,
+ * The operation contains an error.
+ * ERROR = 2
+ */
+
+const DEPRECATED_STATUS_CODE_OK = 0;
+const STATUS_UNSET = 0;
+const STATUS_OK = 1;
+// const STATUS_ERROR = 2;
+
+class InstanaExporter {
+  /**
+   * @param {{ agentKey?: string, endpointUrl?: string }}
+   */
+  constructor({ agentKey, endpointUrl }) {
+    this._isShutdown = false;
+    // If endpoint URL and agent key are not provided, we don't set them in an attempt to use env vars, if they were set
+    if (endpointUrl) {
+      process.env[instanaEndpointUrlEnvVar] = endpointUrl;
+    }
+    if (agentKey) {
+      process.env[instanaAgentKeyEnvVar] = agentKey;
+    }
+
+    if (!process.env[instanaEndpointUrlEnvVar] || !process.env[instanaAgentKeyEnvVar]) {
+      throw new Error('The Instana endpoint URL and/or the agent key were not provided');
+    }
+
+    instanaEnvironment.validate();
+    instanaBackendConnector.init();
+  }
+
+  /**
+   * @param {Array.<ReadableSpan>} spans
+   * @param {(result: import('@opentelemetry/core').ExportResult) => void} resultCallback
+   */
+  export(spans, resultCallback) {
+    const validSpans = spans.map(this._transform.bind(this));
+
+    if (!this._isShutdown) {
+      this._sendSpans(validSpans, resultCallback);
+    }
+  }
+
+  /**
+   * @returns {Promise.<void>}
+   */
+  shutdown() {
+    /**
+     * In the future, there may be the need to "end" any kind of process or network communications with our backend.
+     * Right now, there isn't seem to be any situation where we need that, as we simply make HTTP requests without
+     * actually making any type of stateful or persistent connections.
+     */
+    this._isShutdown = true;
+    return Promise.resolve();
+  }
+
+  /**
+   * @param {Array.<InstanaBaseSpan>} spans
+   * @param {(result: import('@opentelemetry/core').ExportResult) => void} resultCallback
+   */
+  _sendSpans(spans, resultCallback) {
+    instanaBackendConnector.sendSpans(spans, (err /* , _responseBody */) => {
+      /** @type {import('@opentelemetry/core').ExportResult} */
+      let result = { code: ExportResultCode.SUCCESS };
+      if (err) {
+        result = { code: ExportResultCode.FAILED, error: err };
+      }
+      resultCallback(result);
+    });
+  }
+
+  /**
+   * @param {ReadableSpan} span
+   * @returns {InstanaBaseSpan}
+   */
+  _transform(span) {
+    /** @type {import('@instana/core/src/tracing/cls').InstanaBaseSpan} */
+    const result = {
+      n: 'otel',
+      f: {
+        e: process.pid
+      },
+      data: {}
+    };
+
+    const traceId = this._buildTraceOrSpanId(span.spanContext().traceId, true);
+
+    if (traceId.length === 32) {
```

Comment:
In reality, there should only be lengths of 16 or 32, but we still could check `(traceId.length >= 16)` instead of `(traceId.length === 32)` to be extra sure.

Reply:
So, this bit of code was pretty much copied from the Java code, and actually the traceId is always 32 hex strings long (see `_buildTraceOrSpanId`).
This verification could actually be removed, since it is guaranteed that the traceId will be 32 hex strings no matter what.

Does this make sense?
^ @basti1302

Reply:
Ah, I didn't see that. Makes sense, yes. 👍